enhancements/kube-apiserver/startup-monitor.md

```diff
@@ -0,0 +1,234 @@
+5. check the revision annotation is the expected one. Set Reason=**NotReady** if this is red and the monitor times out.
+6. checking status.ready. Set Reason=**NotReady** if this is red and the monitor times out.
+
+The timeout for the readiness check is defined to be 5 min for the following reasons:
+
+- this timeout is for the disaster recovery case, not for the happy case. Hence, we don't have to get into the range of 60s of maxmimal API outage, i.e. around 45s as upper bound for the happy case readiness time.
+- 5 min is already a very generous duration for kube-apiserver startup. In normal situation we are in the area of O(30s).
+
+The timeout is to be specified in the operator. Hence, another operator might want different timeouts.
+
+### Retries
+
+The readiness procedure above has the risk of considering external reasons as fatal reasons for kube-apiserver readiness. E.g. etcd unstable or kubelet not updating the mirror pod. Both would led us to believe the new revision is broken.
+This is different in HA OpenShift because there is no time limit exists for a new static pod to start up. It can keep crash-looping for hours and eventually start-up to finish the roll-out.
+
+In order to mitigate this difference in error behaviour, a failing revision in SNO will be retried to roll-out after >=N minutes, N=O(10), a reasonable time for the cluster to recover from a previous attempt. In the time in-between, the fallback mechanism will move the cluster back to the old revision, until the retry is started.
+
+We will start with infinite retries, but use an exponential backoff starting at N=O(10) as described above, moving up to 6h maximum (i.e. around 5 min API outage per 6h).
+
+While retries are attempted, the operator will report Degraded=true.
+
+Retries are not attempted for the following reasons:
+
+- CrashLooping
+- NeverStartedUp
+
+Note that in case of EtcdUnhealthy we fall back (because e.g. the etcd endpoints got wrong), but we also retry because we want to roll out the configuration changes (e.g. cert rotations). Clearly, there is also the case when etcd is just down, and in this case the retry wouldn't help to solve the problem. But eventually the cluster will recover when the etcd outage is solved.
+
+### Coordination of Installer and Startup-Monitor
+
+Startup-monitor and installer to non-atomic work, and hence we need to coordinate work between them to protect the following case:
+an installer is in progress and wants to install a new revision, the current revision is not ready and we are about to fall back to the previous version. The installer writes the new file and we immediately overwrite it.
+
+Hence, we will write a lock file to /var/lock/kube-apiserver-installer.lock, and make the installer and the startup-monitor readiness procedure to wait for it.
+
+### Pruner
+
+The pruner today does not know about the `last-known-good` symlink to the last successful revision. It must be extended to ignore it and the linked revision.
```

Comment:
would it be okay to extend the pruner controller to read the annotations from the kas.pod and put found revision onto a protected list before starting the pruner pod?